### PR TITLE
Config: injection verification during the native image build

### PR DESCRIPTION
**Pull request by mkouba**
- verify the current ImageMode when a config object is injected
- fail if injected  during the static initialization phase of a native image build

See also discussion in https://github.com/quarkusio/quarkus/discussions/36125

The error message looks like:
```
========================================================================================================================
POSSIBLE CONFIG INJECTION PROBLEM DETECTED
------------------------------------------------------------------------------------------------------------------------
A config object was injected during the static initialization phase of a native image build.
This may result in unexpected errors.
The injected value was obtained at native image build time and cannot be updated at runtime.

Injection point: org.acme.config.Failure#init()
Solutions:
	- If that's intentional then annotate the injected field/parameter with @io.quarkus.arc.config.NativeBuildTime to eliminate the false positive
	- You can leverage the programmatic lookup to delay the injection of a config property; for example '@ConfigProperty(name = "foo") Instance<String> foo'
	- You can try to use a normal CDI scope to initialize the bean lazily; this may help if the is only injected but not directly used during the static initialization phase
========================================================================================================================

```

### CODE REVIEW

File: extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/NativeBuildConfigCheckInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                    if (qualifier instanceof ConfigProperty) {
+                        propertyName = ((ConfigProperty) qualifier).name();
+                    }
+                }
+                if (propertyName != null && buildAndRunTimeFixed.contains(propertyName)) {
+                    return;
+                }
+            }
+        }
+        if (ImageMode.current() == ImageMode.NATIVE_BUILD) {
```
Comment (yrodiere) [inline extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/NativeBuildConfigCheckInterceptor.java:64]: Would it make sense to put this check at the very top, to avoid unnecessary computation at (native) runtime?

Or maybe you could add a GraalVM substitution to make this method a no-op at native runtime?

File: extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/ConfigMappingCreator.java [hunk 2]
```diff
@@ -23,6 +24,10 @@ public Object create(SyntheticCreationalContext<Object> context) {
             throw new IllegalStateException("No current injection point found");
         }
 
+        if ((boolean) context.getParams().get("nativeBuild")) {
+            NativeBuildConfigCheckInterceptor.verifyCurrentImageMode(Set.of());
```
Comment (yrodiere) [inline extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/ConfigMappingCreator.java:28]: IIRC the Set is empty here because we assume none of the properties being injected are "build and runtime fixed"... but can't a config mapping be completely "build and runtime fixed"? If so we'd want to skip this check, no?

File: extensions/arc/runtime/src/main/java/io/quarkus/arc/config/NativeBuildTime.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+/**
+ * A config property injected during the static initialization phase of a native image build may result in unexpected errors
```
Comment (yrodiere) [inline extensions/arc/runtime/src/main/java/io/quarkus/arc/config/NativeBuildTime.java:11]:
```suggestion
 * Allows a configuration property injection point to retrieve values during static initialization.
 * <p>
 * A config property injected during the static initialization phase of a native image build may result in unexpected errors
```

File: extensions/arc/deployment/src/test/java/io/quarkus/arc/test/config/nativebuild/NativeBuildConfigInjectionOkTest.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    @RegisterExtension
+    static final QuarkusProdModeTest TEST = new QuarkusProdModeTest()
+            .setBuildNative(true)
+            .withApplicationRoot(
+                    root -> root.addClass(Ok.class));
+
+    @Test
+    void test() {
+    }
+}
```
Comment (yrodiere) [inline extensions/arc/deployment/src/test/java/io/quarkus/arc/test/config/nativebuild/NativeBuildConfigInjectionOkTest.java:19]: I suppose you didn't add a test for "runtime-fixed" configuration properties because there is none available in the "Arc" extension?

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java [hunk 5]
```diff
@@ -474,67 +473,6 @@ void validateConfigPropertiesInjectionPoints(
...
-                        if (annotationInstances != null && annotationInstances.size() == 1) {
-                            AnnotationInstance annotationInstance = annotationInstances.get(0);
-                            if (annotationInstance.target().equals(declaringClass)) {
-                                LOGGER.warn("Directly injecting a " +
-                                        injectionTypeName +
-                                        " into a " +
-                                        declaringClassCandidate +
-                                        " may lead to unexpected results. To ensure proper results, please " +
-                                        "change the type of the field to " +
-                                        ParameterizedType.create(DotNames.INSTANCE, new Type[] { type }, null) +
```
Comment (yrodiere) [inline extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java:524]: I see this earlier version of the check took into account the type being injected in the error message (i.e. it would suggest `Instance<MyConfig>` instead of `Instance<String>`, IIUC).

Would it make sense to still do that with the new approach, or is it too complex to be worth it?

**Review by yrodiere — COMMENTED** [review]
Thanks a lot! It looks like a nice improvement. I added a few comments below, hopefully not all of them are silly questions :)

**Comment by radcortez** [comment]
If we remove the JVM mode check, you may still get unexpected results. The issue is that the available `Config` differs between `STATIC_INIT` and `RUNTIME`, even in JVM mode.  For instance, sources like Vault, or Database, are not available during `STATIC_INIT`, meaning that they can override a configuration for components initialized statically.

**Comment by yrodiere** [comment]
> If we remove the JVM mode check, you may still get unexpected results. The issue is that the available `Config` differs between `STATIC_INIT` and `RUNTIME`, even in JVM mode. For instance, sources like Vault, or Database, are not available during `STATIC_INIT`, meaning that they can override a configuration for components initialized statically.

So you think the check should happen in JVM mode too, and the annotation should be renamed to something like `@StaticInitialization`, and the error message rephrased accordingly?

**Comment by radcortez** [comment]
> But in JVM mode, static init config is apparently influenced by what developers consider "runtime" config (e.g. env variables or system properties I suppose, or perhaps config files on the filesystem). At least that seems to be what triggered all the fuss in #36125 . If that's not the case, then I don't understand #36125 anymore, because the problem experienced in native mode should have happened in JVM mode as well.

Correct. Both JVM and Native go through a `STATIC_INIT` config phase, but each can have different configuration. For instance, JVM mode can add configuration found in `config` folder, which is not available during the native image build. This does make things super confusing.

> In that light, the fact that Vault or Database config sources don't affect static init config in JVM mode is... confusing? I suppose there are technical reasons (e.g. the DB connection pool is not available during static init), but it's definitely not just "the same".

Correct, DB connection pool is not available during `STATIC_INIT`. I guess it would be possible to have Vault available during JVM `STATIC_INIT`, but not during native image build (while technically possible, I guess that you don't want to have Vault config participate in that case, because it needs an external URL that may not be available in the build machine). This would require different registration of sources depending on the phase and the target, which would make things even more confusing.

>   2. In JVM mode, that's more complicated. If we can identify sources that may be problematic (vault, database, ...), then that's great and we should have the check trigger a failure, but by default it probably shouldn't.

What we can probably do, is record the injected configuration in `STATIC_INIT` components and match the values after the runtime configuration is available. If you find a mismatch, it means that a runtime `ConfigSource` made an override to the injected value.

**Comment by yrodiere** [comment]
> What we can probably do, is record the injected configuration in `STATIC_INIT` components and match the values after the runtime configuration is available. If you find a mismatch, it means that a runtime `ConfigSource` made an override to the injected value.

That's interesting... And I think there is such a feature somewhere already, about build time config being overridden later, right?

@mkouba WDYT?

**Comment by radcortez** [comment]
> That's interesting... And I think there is such a feature somewhere already, about build time config being overridden later, right?

Yes, we do it here: https://github.com/quarkusio/quarkus/blob/ebe995ab7c5473b02db9d8afeeb50c6c595bb15c/core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigRecorder.java#L28-L76, but I believe that for this discussed case it would be easier to implement.

For build time, we generate a MAX ordinal source to prevent users to override the values, but we have to discard it to search for values that may have been overridden in other sources by matching the recorded build time time properties against the runtime properties.

File: extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/NativeBuildConfigCheckInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            AnnotatedParameter<?> param = (AnnotatedParameter<?>) annotated;
+            if (param.getDeclaringCallable() instanceof AnnotatedConstructor) {
+                return param.getDeclaringCallable().getDeclaringType().getJavaClass().getName() + "()";
+            } else {
+                return param.getDeclaringCallable().getDeclaringType().getJavaClass().getName() + "#"
+                        + param.getDeclaringCallable().getJavaMember().getName() + "()";
+            }
+        }
+        return injectionPoint.toString();
+    }
```
[the remark this answers is not in the document]
Comment (mkouba) [reply extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/NativeBuildConfigCheckInterceptor.java:112]: AFAIK we do have something in build steps but this is runtime... :shrug:

File: extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/NativeBuildConfigCheckInterceptor.java (hunk 1 above)
Comment (mkouba) [reply extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/NativeBuildConfigCheckInterceptor.java:64]: I will put it at the very top. Although I don't think it will be a measurable improvement ;-)

> Or maybe you could add a GraalVM substitution to make this method a no-op at native runtime?

I don't think it's worth the effort.

File: extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/ConfigMappingCreator.java (hunk 2 above)
Comment (mkouba) [reply extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/ConfigMappingCreator.java:28]: Yes, a `ConfigRoot` defined as a config mapping is something I need to check.

Because we do register a synthetic bean for any `BUILD_AND_RUN_TIME_FIXED`/`RUN_TIME` config root [here](https://github.com/quarkusio/quarkus/blob/main/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java#L238-L257)  and my guess would be that an injected config root with a config mapping would result in an ambiguous dependency error...

File: extensions/arc/runtime/src/main/java/io/quarkus/arc/config/NativeBuildTime.java (hunk 3 above)
Comment (mkouba) [reply extensions/arc/runtime/src/main/java/io/quarkus/arc/config/NativeBuildTime.java:11]: We should improve the wording here, as it's not only about injecting properties.

File: extensions/arc/deployment/src/test/java/io/quarkus/arc/test/config/nativebuild/NativeBuildConfigInjectionOkTest.java (hunk 4 above)
Comment (mkouba) [reply extensions/arc/deployment/src/test/java/io/quarkus/arc/test/config/nativebuild/NativeBuildConfigInjectionOkTest.java:19]: Yes.

File: extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/ConfigMappingCreator.java (hunk 2 above)
Comment (mkouba) [reply extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/ConfigMappingCreator.java:28]: FYI it's not a problem because a config root implemented as a config mapping is not considered a `RootDefinition`...

**Comment by mkouba** [comment]
> > That's interesting... And I think there is such a feature somewhere already, about build time config being overridden later, right?
> 
> Yes, we do it here: ...

By the way, this results in a relatively large generated `StartupTask` class as it contains all build-time and build-runtime-fixed values, ~ 25KB for the config quickstart.

In theory, we could record the values for all `@ConfigProperty` injection points and check those values during `RUNTIME_INIT` (and fail if a mismatch occurs), and handle the config mapping injection points differently... not sure how though :-).

**Comment by radcortez** [comment]
> By the way, this results in a relatively large generated `StartupTask` class as it contains all build-time and build-runtime-fixed values, ~ 25KB for the config quickstart.

We shouldn't need this. Only the properties that were injected in the static init classes. If we do need them, they are already recorded in a `ConfigSource` set with `Integer.MAX`, so querying the `Config` system will always return the recorded expected value

> In theory, we could record the values for all `@ConfigProperty` injection points and check those values during `RUNTIME_INIT` (and fail if a mismatch occurs), and handle the config mapping injection points differently... not sure how though :-).

Config mappings are not registered by default for `STATIC_INIT`, you need to explicitly annotate the mapping with `@StaticInitSafe`. I think we can consider that if you add that annotation you are on your own :)

**Comment by mkouba** [comment]
> > In theory, we could record the values for all `@ConfigProperty` injection points and check those values during `RUNTIME_INIT` (and fail if a mismatch occurs), and handle the config mapping injection points differently... not sure how though :-).
> 
> Config mappings are not registered by default for `STATIC_INIT`, you need to explicitly annotate the mapping with `@StaticInitSafe`. I think we can consider that if you add that annotation you are on your own :)

Yes, that's actually good news. So injecting a config mapping that is not annotated with `@StaticInitSafe` results in a failure during `STATIC_INIT`. The same applies to `RUN_TIME` config roots. And `BUILD_AND_RUN_TIME_FIXED` config roots shouldn't change. So if we ignore the programmatic API for now (which is IMO ok) we could just record the properties injected during `STATIC_INIT`, compare the runtime values and fail the `RUNTIME_INIT` if a mismatch is found.

In other words, if a CDI bean that is created during `STATIC_INIT` injects a config property we will record the injected value and then during `RUNTIME_INIT` we'll compare the injected values with the "final" values. If a mismatch is found then we will fail with an actionable error.

Does it make sense? @radcortez @yrodiere

**Comment by yrodiere** [comment]
> Does it make sense? @radcortez @yrodiere

It does to me.

The one downside compared to your current solution is that application developers might end up in a situation where their config gets injected without error, and they end up using it during static init, triggering other, less explicit errors down the line because the property has no value during static init.

But I suppose it has more to do with other libraries (e.g. Hibernate Validator) at this point, and at least in such a scenario the developer will get a stacktrace showing them where the problem happens, 

> So if we ignore the programmatic API for now (which is IMO ok) 

By "programmatic API", do you mean the use of `Instance` (e.g. `@ConfigProperty(...) Instance<String>`), or retrieval of config properties via `Arc.container().instance(...)` (is that even possible?), or retrieval of config properties through Smallrye's `ConfigProvider.getConfig()`? I'm not sure I'd consider any of those negligible...

**Comment by mkouba** [comment]
> The one downside compared to your current solution is that application developers might end up in a situation where their config gets injected without error, and they end up using it during static init, triggering other, less explicit errors down the line because the property has no value during static init.

Yes, on the other hand we would not fail if the injected value did not change which is good ;-)

> > So if we ignore the programmatic API for now (which is IMO ok)
> 
> By "programmatic API", do you mean the use of `Instance` (e.g. `@ConfigProperty(...) Instance<String>`), or retrieval of config properties via `Arc.container().instance(...)` (is that even possible?), or retrieval of config properties through Smallrye's `ConfigProvider.getConfig()`? I'm not sure I'd consider any of those negligible...

I mean the _"retrieval of config properties through Smallrye's `ConfigProvider.getConfig()`"_. `@ConfigProperty(...) Instance<String>` and `Arc.container().instance(...)` would be handled (since the CDI beans are obtained under the hood).

**Comment by radcortez** [comment]
> I mean the _"retrieval of config properties through Smallrye's `ConfigProvider.getConfig()`"_. `@ConfigProperty(...) Instance<String>` and `Arc.container().instance(...)` would be handled (since the CDI beans are obtained under the hood).

Correct.

> I mean the _"retrieval of config properties through Smallrye's `ConfigProvider.getConfig()`"_. `@ConfigProperty(...) Instance<String>` and `Arc.container().instance(...)` would be handled (since the CDI beans are obtained under the hood).

`ConfigProvider.getConfig` will just give you the `Config` instance that is available at each phase. We register a `Config` instance for `STATIC_INIT` and that is the one you get during that phase. When we move to runtime, we release the old instance and register a new instance with the components now available for runtime too.

**Comment by mkouba** [comment]
This PR was superseded by https://github.com/quarkusio/quarkus/pull/36281.
